Dev

mikaelmessias opened the pull request.

mikaelmessias committed:
Fix conflit on the bfs_tree_remove

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

mikaelmessias committed:
Added .gitignore rule for VSCode files

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

mikaelmessias committed:
Added a macro for blue ANSI color escaping sequence

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

mikaelmessias committed:
Fixed BFS Tree implementation

	Also, the BFS cmdline is fully working now. Christmas miracles
are real!

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

mikaelmessias committed:
Merge branch 'test' into dev

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

rafaelcn commented on src/bfs_tree.c:
```diff
@@ -26,29 +26,110 @@
 #include "bfs_errors.h"
+#include "bfs_macros.h"
+
+#define newline printf("\n")
```
I think by the excessive usage of this macro it has to be moved to `bfs_macros.h`.

mikaelmessias replied:
Got it.

rafaelcn commented on src/bfs_macros.h:
```diff
@@ -24,12 +24,17 @@
+#define ANSI_COLOR_BLUE     "\x1b[34;1m"
 #define ANSI_COLOR_RESET   "\x1b[0m"
+
+// ANSI font effects escaping sequences
```
I think that the old way was better. Not it is somewhat redundant.

mikaelmessias replied:
I was going to use this space to create macros for font effects (like bold) but I ended up giving up on the idea and just forgot to erase the comment.

rafaelcn replied:
That's okay!

rafaelcn commented on src/bfs_cmdline.c:
```diff
@@ -189,11 +304,19 @@ void cmdline_start(BFSTree *root) {
+            // FIXME: Fix size of the option buffer
+            sprintf(ebuffer, "Given command \"%s\" is invalid", option);
+
+            // bfs_pferror(stderr, ebuffer, __LINE__, __FILE__, BFS_WARNING);
```
You may remove this as it isn't a invalid command.

rafaelcn commented on src/bfs_tree.c:
```diff
@@ -26,29 +26,110 @@
+        return EMPTY_NODE;
+    }    
+    
+    // 1º: verifica a quantidade de níveis
```
Describe everything in english please.

mikaelmessias replied:
Fixed.

mikaelmessias committed:
Fixed the remove function to show appropriate error messages.

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

mikaelmessias committed:
Added two functions to bfs_cmdline

	Since both cd and search functions were getting too big, I realized it would be better to move them to their own functions to make the code cleaner.

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

rafaelcn commented on src/bfs_tree.c:
```diff
@@ -77,106 +174,122 @@ BFSNode *bfs_tree_create_node(char fname[BFS_MAX_NAME_LENGTH]) {
+    return bfs_tree_search_child(child->next, s);
+}
+
+BFSNode *tmp;
```
Why there's a global temporary variable over here?

rafaelcn replied:
A better solution is needed before this gets merged in the `dev` :-)

mikaelmessias replied:
Fixed.

mikaelmessias replied:
Fixed.

rafaelcn commented on src/bfs_tree.c:
```diff
@@ -77,106 +174,122 @@ BFSNode *bfs_tree_create_node(char fname[BFS_MAX_NAME_LENGTH]) {
-    if(root == NULL) {
-        return 0;
+    else if(strcmp(arg, "-h") == 0) {
+        goto help;
```
Although I don't usually prohibit this kind of thing, this in this case is not any useful. Can you please create a function to show specific help inside BFS?

mikaelmessias replied:
Fixed.

rafaelcn commented on src/bfs_tree.c:
```diff
@@ -187,73 +300,36 @@ int bfs_tree_remove(BFSTree root, char path[BFS_PATH_SIZE]) {
     }
```
Looks like we have some space over here. Can you please remove it?

rafaelcn commented on src/bfs_tree.h:
```diff
@@ -30,22 +30,29 @@
 
-#define EMPTY_TREE (BFSTree *)0
-#define EMPTY_NODE (BFSNode *)0
+#define EMPTY_TREE (void *)0
```
Explain me the reason behind this change. I don't see any reason to change it, as it is a null `BFSTree` node instead of a `void` one.

mikaelmessias replied:
I did this because I thought it would be easier to understand. For example, in some functions that return BFSTree type, we sometimes check if a node is null instead of the tree itself. Like here:

```c
BFSTree *bfs_tree_insert(BFSTree *root, char path[BFS_PATH_SIZE]) {
    if(check_pointer(root)) {
        return EMPTY_TREE;
    }
    else if(check_pointer(*root)) {
        return EMPTY_TREE;
    }
    ...
}
```
Without this change, we can't return EMPTY_NODE in a BFSTree type function or EMPTY_TREE in a BFSNode type function, like this:

```c
BFSTree *bfs_tree_insert(BFSTree *root, char path[BFS_PATH_SIZE]) {
    if(check_pointer(root)) {
        return EMPTY_TREE;
    }
    else if(check_pointer(*root)) {
        return EMPTY_NODE;
    }
    ...
}
```

I know it's not very important, but I also think it does not make much sense to return EMPTY_TREE instead of EMPTY_NODE when we're checking a node instead of the whole tree or the opposite. But I understand that it is not quit right to do it this way, so I am returning to the old pattern.

rafaelcn commented on src/bfs_tree.h:
```diff
@@ -30,22 +30,29 @@
-#define EMPTY_TREE (BFSTree *)0
-#define EMPTY_NODE (BFSNode *)0
+#define EMPTY_TREE (void *)0
+#define EMPTY_NODE (void *)0
```
Same applies here.

rafaelcn commented on src/bfs_tree.c:
```diff
@@ -77,106 +174,122 @@ BFSNode *bfs_tree_create_node(char fname[BFS_MAX_NAME_LENGTH]) {
+        return;
+    }
+    else if(root->child == NULL && root->fpermissions == BFS_NODE_IS_DIR) {
+        bfs_pferror(stderr, "Empty directory", __LINE__, __FILE__,
```
This shouldn't be printed as a warning. We certainly have to open an issue for this.

rafaelcn commented on src/bfs_tree.c:
```diff
@@ -59,14 +140,30 @@ BFSNode *bfs_tree_create_node(char fname[BFS_MAX_NAME_LENGTH]) {
+    return node;
+}
 
+BFSNode *bfs_tree_create_file(char fname[BFS_MAX_NAME_LENGTH]) {
```
Awesome decision :+1:

mikaelmessias replied:
Thanks!

rafaelcn reviewed (changes requested):
Thanks for the improvements! It is a great change and there are some changes we need to do in order to merge this stuff and then send it to the master :-)

rafaelcn commented:
Hello @mikaelmessias, how is Mars over there?

mikaelmessias committed:
Moved the newline macro to bfs_macros.h

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

mikaelmessias committed:
Improved bfs_tree_get_child function

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

mikaelmessias committed:
Improved how to show specific help in bfs_tree_print function

	Some minor changes were made as requested, plus a new color
escaping sequence and a small change in the cmdline to display the
current path with this color.

Signed-off-by: Mikael Messias <mikaelmessias@outlook.com>

rafaelcn commented:
I think we've got some conflicts to solve
